pyccel/codegen/printing/ccode.py

```diff
@@ -722,11 +722,32 @@ def _print_IndexedElement(self, expr):
                         inds[i] = Slice(ind, PyccelAdd(ind, LiteralInteger(1)), LiteralInteger(1))
                 inds = [self._print(i) for i in inds]
                 return "array_slicing(%s, %s, %s)" % (base_name, expr.rank, ", ".join(inds))
-            inds = [self._print(i) for i in inds]
+            inds = [self._cast_to(i, NativeInteger(), 8) for i in inds]
         else:
             raise NotImplementedError(expr)
         return "%s.%s[get_index(%s, %s)]" % (base_name, dtype, base_name, ", ".join(inds))
 
+    def _cast_to(self, expr, dtype, precision):
+        """ add cast to an expresion
+        parameters
+        ----------
+            expr :
+                the expression to be casted
+            dtype : Datatype
+                base type of the cast
+            precision : integer
+                precision of the base type of the cast
+
+        Return
+        ------
+            String
+                Return string that contains the expression casted to the desired type
+        """
+        if (expr.dtype != dtype or expr.precision != precision):
+            cast=self.find_in_dtype_registry(self._print(expr.dtype), precision)
+            expr = self._print(expr)
+            return '({cast}){expr}'.format(cast=cast, expr=expr)
+        return self._print(expr)
```

Comment:
I see that this method doesn't always add cast the expression like what you explain in the docstring.
can you update the docstring desc?
before that, I see that this function serves only this [line-725](https://github.com/pyccel/pyccel/pull/714/files#diff-f1038cb7408f08867011c5feb494cd9e7a5a435a3b6f6438ef6cd3bdf3c90fbeR725), can you explain more why we need this function, and where it can help us.

Reply:
having a function that does the casting is better than having duplicated code.

Reply:
In lots of the places where we need to cast we cannot use `_print(expr)` (e.g. when casting `floor`, calling `_print(expr)` would cause an infinite loop).
I wonder if we could make this more general, eg by returning a format string

Reply:
Does making the function take the expr string, type, and precision as arguments fixing this? because passing the AstNode will result in an infinite loop.

Reply:
Passing the `AstNode` only leads to an infinite loop if `_print(expr)` is called. Making it take a string in place of `expr` would fix this, but you can no longer do the important test `if (expr.dtype != dtype or expr.precision != precision):`
So you would have to pass `expr` and `str_expr`.

Reply:
is there a way to do it without passing `expr` and `expr_str` at the same time? i can't think of a way.

Reply:
As I mentioned above, this could be done by returning a format string. I.e:
```
if ...
    ...
    return '({cast}){{}}'
else:
    return '{}'

But you still need to manage both in the calling function

Reply:
wouldn't this make it easier to forget to add to the code.

Reply:
even if passing the `expr` and `expr_str` is a little bit odd but it will make it easier to not make a mistake.

Reply:
Hmm, I hadn't considered that. Possibly. Would it not raise an error in C? Or will it default to 0?
@rakati do you have any suggestions?

Reply:
You could make `expr_str` an optional parameter which defaults to `self._print(expr)`

Reply:
will this not the same as calling `self._print(expr)` inside the function?

Reply:
with returning the format i think its oket unless you want to use `_cast_to` in a complex format string as the line will be too long a call to `_cast_to` will be like `self._cast_to(i, NativeInteger(), 8).format(self._print(i))`

Reply:
If it is misused then yes, but this will easily be spotted. In cases where an infinite loop would occur, we would need to make sure that we provided `expr_str`

Reply:
> with returning the format I think it's okay unless you want to use _cast_to in a complex format string as the line will be too long.

You can always split the line if it becomes too long by saving the format string in a variable. In any case passing it as an argument will give almost the same line length

Reply:
which one seems more intuitive?

Reply:
Personally I don't have a preference

Reply:
can we even use `self._print` as optional argument ??
i got self is not defined error.

Reply:
```
def _cast_to(self, expr, dtype, precision, expr_str = None):
    if expr_str is None:
        expr_str = self._print(expr)
```